### ISSUE TITLE
Support markers with line plots

### ISSUE THREAD

**Issue by GregVernon** [racket/plot#118]
Various plotting tools (Excel, Matplotlib, Matlab, etc) support the use of markers in a line plot, which can then be represented in the legend.  Typically these tools support markers on both symbolic and discrete data plots.  For example, consider the Matlab code below:

```matlab
clear
close all

syms x
f(x) = sin(x)
N = 18
x_data = linspace( 0, 2*pi, N );
y_data = sin(x_data) + ( rand(1,N) - 0.5 ) / 5;

figure
hold on
fplot( f, [0, 2*pi], LineWidth=2, LineStyle="-", Marker="o", MarkerFaceColor="w", MarkerEdgeColor="k", DisplayName="Exact" )
plot( x_data, y_data, LineWidth=2, LineStyle="-", Marker="s", MarkerFaceColor="w", MarkerEdgeColor="k", DisplayName="Test Data" )
legend()
```

Which produces the below plot, note that the `Exact` plot is a symbolic function while the `Test Data` is discrete.
![image](https://github.com/racket/plot/assets/868479/f65d4909-4183-47a6-bea3-18f868854443)

Essentially, it would be nice if I could make this Racket plot replicate the behavior shown above:

```racket
#lang racket/base
(require racket/stream)
(require (only-in racket/list range) )
(require plot)
(require math)


(define (randsin x)(+ (sin x) (/ ( - (random) 0.5) 5) ) )
(define xdata (range 0 (* 2 pi) (/ pi 9 ) ) )
(define ydata (map (lambda (x) (randsin x) ) xdata ) )

(parameterize ([plot-width    400]
               [plot-height   400]
               [plot-x-label  #f]
               [plot-y-label  #f])
  (plot (list
         (function sin 0 (* 2 pi) #:label "exact" )
         (lines (map vector xdata ydata ) #:label "discrete" ) )
  )
)
```

![image](https://github.com/racket/plot/assets/868479/f2111309-0d76-4835-bfff-24cbbf9ea866)

**Comment by alex-hhh** [comment]
You can already do this with the Racket plot package, but this is done by composing existing renderers, rather than having an extra argument to renderers.  For example, you can define `line-with-markers` and `function-with-markers` helpers to return both a lines and a points renderers, but they can be used as a single renderer in the `plot` call:

```scheme
#lang racket

(require (only-in racket/list range) )
(require plot)
(require math)

(point-size 10)
(line-width 2)
(point-line-width 1.5)
(plot-pen-color-map 'set1)

(define (lines-with-markers
         data
         #:label [label #f]
         #:sym [sym (point-sym)]
         #:color [color 1]
         #:sym-color [sym-color 0])
  (list
   (lines data #:label label #:color color)
   (points data #:sym sym #:color sym-color)))

(define (function-with-markers
         fn min max
         #:label [label #f]
         #:sym [sym (point-sym)]
         #:samples [samples 20]
         #:color [color 1]
         #:sym-color [sym-color 0])
  (define step (/ (- max min) samples))
  (define data
    (for/list ([x (in-inclusive-range min max step)])
      (vector x (fn x))))
  (list
   (function fn min max #:label label #:color color)
   (points data #:sym sym #:color sym-color)))

(define (randsin x)(+ (sin x) (/ ( - (random) 0.5) 5) ) )
(define xdata (range 0 (* 2 pi) (/ pi 9 ) ) )
(define ydata (map (lambda (x) (randsin x) ) xdata ) )

(parameterize ([plot-width    400]
               [plot-height   400]
               [plot-x-label  #f]
               [plot-y-label  #f])
  (plot (list
         (function-with-markers sin 0 (* 2 pi)
                                #:label "exact"
                                #:sym 'circle3
                                #:samples (length ydata)
                                #:color 0
                                #:sym-color "black")
         (lines-with-markers (map vector xdata ydata )
                             #:label "discrete"
                             #:sym 'square
                             #:color 1
                             #:sym-color "black"))
        #:legend-anchor 'top-right))

```

![pr-118-a](https://github.com/racket/plot/assets/11592690/65881b90-35a8-4164-b23c-e891347fe97f)

You'll notice that the legend entries don't contain the markers, and fixing that requires some more in-depth knowledge, but here is how you can overlay legend keys onto each other from two plots:

```racket
#lang racket

(require racket/stream)
(require (only-in racket/list range) )
(require plot)
(require math)

(require plot/private/common/types)     ; legend-entry

(plot-new-window? #t)
(point-size 10)
(line-width 2)
(point-line-width 1.5)
(plot-pen-color-map 'set1)

(define (overlay-legend-entries r1 r2)
  ;; NOTE: label functions can return "treeof" legend entries -- we don't
  ;; handle that here as `function`, `lines`, and `points` renderers don't do
  ;; that.
  (define l1 (renderer2d-label r1))
  (define l2 (renderer2d-label r2))
  (lambda (rect)
    (define e1 (l1 rect))
    (define e2 (l2 rect))
    (legend-entry
     ;; Take the label from the first entry
     (legend-entry-label e1)
     ;; And the draw functions draws both entries in the same place (on top of
     ;; each other)
     (lambda (pd w h)
       ((legend-entry-draw e1) pd w h)
       ((legend-entry-draw e2) pd w h)))))

(define (lines-with-markers data
                            #:label [label #f]
                            #:sym [sym (point-sym)]
                            #:color [color 1]
                            #:sym-color [sym-color 0])
  (define l (lines data #:label label #:color color))
  (define p (points data #:sym sym #:color sym-color #:label ""))
  (list
   (struct-copy renderer2d l [label (overlay-legend-entries l p)])
   (struct-copy renderer2d p [label #f])))

(define (function-with-markers fn min max
                               #:label [label #f]
                               #:sym [sym (point-sym)]
                               #:samples [samples 20]
                               #:color [color 1]
                               #:sym-color [sym-color 0])
  (define step (/ (- max min) samples))
  (define data
    (for/list ([x (in-inclusive-range min max step)])
      (vector x (fn x))))
  (define f (function fn min max #:label label #:color color))
  (define p (points data #:sym sym #:color sym-color #:label ""))
  (list
   (struct-copy renderer2d f [label (overlay-legend-entries f p)])
   (struct-copy renderer2d p [label #f])))


(define (randsin x)(+ (sin x) (/ ( - (random) 0.5) 5) ) )
(define xdata (range 0 (* 2 pi) (/ pi 9 ) ) )
(define ydata (map (lambda (x) (randsin x) ) xdata ) )


(parameterize ([plot-width    400]
               [plot-height   400]
               [plot-x-label  #f]
               [plot-y-label  #f])
  (plot (list
         (function-with-markers sin 0 (* 2 pi)
                                #:label "exact"
                                #:sym 'circle3
                                #:samples (length ydata)
                                #:color 0
                                #:sym-color "black")
         (lines-with-markers (map vector xdata ydata )
                             #:label "discrete"
                             #:sym 'square
                             #:color 1
                             #:sym-color "black"))
        #:legend-anchor 'top-right
        #:out-file "pt-118-b.png"))
```
![pt-118-b](https://github.com/racket/plot/assets/11592690/73b824b2-e6fa-446d-8883-4ca15e7808a6)

**Comment by GregVernon** [comment]
Thanks!  We'd internally done the compositing of line and scatter plots but hadn't figured out how to create the composed legend entry.  I honestly thought it would be more challenging to get this to work than your example above.  I would propose that the `lines-with-markers` and `function-with-markers` functionality be added to racket/plot as "out-of-the-box" functionality -- whether as new methods or updates to existing `lines` and `function` methods.

**Comment by alex-hhh** [comment]
Hi @GregVernon , I think you are right that it is non-obvious how to add the markers, so I updated the `lines` and `function` renderers to support a `#:marker` option, as well as other options to control the size and color of the markers. This will be available when Racket 8.10 is released, but you can  build the code yourself, see #120 , if you want to try it out earlier.
